qiskit/circuit/_standard_gates_commutations.py

```diff
@@ -0,0 +1,3264 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Dictionary containing commutation relations of unparameterizable standard gates. The dictionary key
+is defined by a pair of gates and, optionally, their relative placement to each other, i.e.:
+
+   ┌───┐┌───┐
+q: ┤ X ├┤ Z ├    -> standard_gates_commutations["x", "z"] = False
+   └───┘└───┘
+
+or
+          ┌───┐
+q_0: ──■──┤ X ├
+     ┌─┴─┐└─┬─┘  -> standard_gates_commutations["cx", "cy"][1, 0] = False
+q_1: ┤ Y ├──■──
+     └───┘
+
+or
+
+q_0: ──■───────
+     ┌─┴─┐
+q_1: ┤ X ├──■──  -> standard_gates_commutations['cx', 'cx'][None, 0] = False
+     └───┘┌─┴─┐
+q_2: ─────┤ X ├
+          └───┘
+Note that the commutation of a pair of gates is only stored in one specific order, i.e. the commutation
+of a pair of cy and cx gates is only stored in standard_gates_commutations["cx", "cy"]. The order of the
+gates is derived from the number of qubits in each gate and the value of the integer representation of
+the gate name.
+
+The values of standard_gates_commutations can be a Boolean value denoting whether the pair of gates
+commute or another dictionary if the relative placement of the pair of gates has an impact on the
+commutation. The relative placement is defined by the gate qubit arrangements as q2^{-1}[q1[i]] where
+q1[i] is the ith qubit of the first gate and q2^{-1}[q] returns the qubit index of qubit q in the second
+gate (possibly 'None'). In the second example, the zeroth qubit of cy overlaps with qubit 1 of cx. Non-
+overlapping qubits are specified as 'None' in the dictionary key, e.g. example 3, there is no overlap
+on the zeroth qubit of the first cx but the zeroth qubit of the second gate overlaps with qubit 1 of the
+first cx.
+"""
+
+standard_gates_commutations = {
```

Comment:
I'm impressed by the amount of checking that this required! But I'm wondering if we could represent (or infer) some of these relations more efficiently than to hardcode them. My main concern here is that a hardcoded table can be challenging to maintain and adding new standard gates comes at quite an overhead. For example
* the `id` clearly commutes with any gate, is this something the compiler should know natively? Right now, adding a new standard gate would require adding an entry here, which is maybe not optimal.
* any diagonal gate commute with controls
* two controlled gates on any number of controls, `C^m A` and `C^n B`, commute if they act on the same and `A` and `B` commute

I might also have missed some discussion on this, sorry if this is the case! 🙂

Reply:
Well, it is my first larger PR so far, so I expected a bit more back and forth. Thanks, @alexanderivrii ! :-) 

I agree that commutation passes should know the commutation of gates such as `id` or gates that do not overlap at all natively. There is some work on the commutation analysis pass as far as I know - maybe this can be addressed in that context.

I agree that some of these relations can be stored more efficiently. However, the lookup of these relations must also be quicker than the corresponding matrix multiplications and another challenge is that these rules should be exhaustive for non-commutation and for commutation. Otherwise, we will need to resort to matrix multiplication again. I guess that we could add a function that is executed before the commutation library is queried to check for these rules but the runtime of the included checks must be offset by their applicability.

Regarding maintenance: ideally you would run something like `build_commutation_library` and `validate_commutation_library` to add a new standard gate. The corresponding diff should only show the new entries.